File: sdi3-14.Cli-REST/src/uo/sdi/clients/rest/restService/RestService.java
package uo.sdi.clients.rest.restService;

import java.util.ArrayList;
import java.util.List;

import javax.ws.rs.ProcessingException;
import javax.ws.rs.client.ClientBuilder;
import javax.ws.rs.client.Entity;
import javax.ws.rs.core.GenericType;
import javax.ws.rs.core.MediaType;

import uo.sdi.clients.rest.dto.Category;
import uo.sdi.clients.rest.dto.Task;

public class RestService {

	private static final String URL_BASE = "http://localhost:8280/sdi3-14.Web/rest/TaskServiceRs";

	private String user;
	private String password;

	private Long userId;

	public RestService(String user, String password) {
		this.user = user;
		this.password = password;
		this.userId = login();
	}

	public Long login() {
		Long u;
		try {
			u = ClientBuilder.newClient()
					.register(new Authenticator(user, password))
					.target(URL_BASE).path("login").request()
					.accept(MediaType.APPLICATION_JSON).get()
					.readEntity(Long.class);
		} catch (javax.ws.rs.ProcessingException p) {
			u = null;
		}
		return u;
	}

	public Task createTask(Task tarea, Long catId) {
		Task t;
		try {
			t = ClientBuilder.newClient()
					.register(new Authenticator(user, password))
					.target(URL_BASE).path("users").path(userId.toString())
					.path("categories").path(catId.toString()).path("tasks")
					.request().accept(MediaType.APPLICATION_XML)
					.post(Entity.entity(tarea, MediaType.APPLICATION_XML))
					.readEntity(Task.class);
		} catch (javax.ws.rs.ProcessingException p) {
			t = null;
		}
		return t;
	}

	public List<Category> findCategoriesByUserID() {
		GenericType<List<Category>> listm = new GenericType<List<Category>>() {
		};

		List<Category> res;

		try {
			res = ClientBuilder.newClient()
					.register(new Authenticator(user, password))
					.target(URL_BASE).path("users").path(userId.toString())
					.path("categories").request()
					.accept(MediaType.APPLICATION_XML).get().readEntity(listm);

		} catch (ProcessingException p) {
			res = new ArrayList<Category>();
		}

		return res;
	}

	public void markTaskAsFinished(Long id) {
		ClientBuilder.newClient().register(new Authenticator(user, password))
				.target(URL_BASE).path("users").path(userId.toString())
				.path("tasks").path(id.toString()).request().put(null);
	}

	public List<Task> getTasksFromCategoryId(Long id) {

		GenericType<List<Task>> listm = new GenericType<List<Task>>() {
		};

		List<Task> res;

		try {
			res = ClientBuilder.newClient()
					.register(new Authenticator(user, password))
					.target(URL_BASE).path("users").path(userId.toString())
					.path("categories").path(id.toString()).path("tasks")
					.request().accept(MediaType.APPLICATION_XML).get()
					.readEntity(listm);
		} catch (ProcessingException p) {
			res = new ArrayList<Task>();
		}

		return res;
	}

}

File: sdi3-14.Cli-REST-CSharp/sdi3-14.Cli-REST-CSharp/restService/RestService.cs
using Newtonsoft.Json;
using sdi3_14.Cli_REST_CSharp.dto;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;

namespace sdi3_14.Cli_REST_CSharp.restService
{
    public class RestService
    {
        private const string URL_BASE = "http://localhost:8280/sdi3-14.Web/rest/TaskServiceRs";


        private String encodedCredentials;
        private Int64? userId;

        public RestService(String user, String password)
        {
            encodedCredentials = "Basic " + System.Convert.ToBase64String(System.Text.Encoding.GetEncoding("ISO-8859-1").GetBytes(user + ":" + password));

        }

        public Int64? login()
        {
            Int64? u = null;

            System.Net.HttpWebRequest webrequest = (HttpWebRequest)System.Net.WebRequest.Create(URL_BASE + "/login");
            webrequest.Method = "GET";
            webrequest.ContentType = "application/json";

            webrequest.Headers.Add("Authorization", encodedCredentials);

            string result;

            try
            {
                using (WebResponse response = webrequest.GetResponse())
                {
                    using (StreamReader reader = new StreamReader(response.GetResponseStream()))
                    {
                        result = reader.ReadToEnd();
                    }
                }

            }
            catch (WebException e)
            {
                return null;
            }


            u = JsonConvert.DeserializeObject<Int64?>(result);

            userId = u;

            return u;
        }

        public dto.Task createTask(dto.Task tarea, long catId)
        {
            string parsedTask = JsonConvert.SerializeObject(tarea);

            dto.Task responseTask = null;

            System.Net.HttpWebRequest webrequest = (HttpWebRequest)System.Net.WebRequest.Create(URL_BASE + "/users/" + userId + "/categories/" + catId + "/tasks");
            webrequest.Method = "POST";
            

            webrequest.Headers.Add("Authorization", encodedCredentials);

            byte[] bytes = Encoding.ASCII.GetBytes(parsedTask);

            var newStream = webrequest.GetRequestStream();
            newStream.Write(bytes, 0, bytes.Length);
            newStream.Close();
            webrequest.ContentType = "application/json";

            string result;

            try
            {
                using (WebResponse response = webrequest.GetResponse())
                {
                    using (StreamReader reader = new StreamReader(response.GetResponseStream()))
                    {
                        result = reader.ReadToEnd();
                    }
                }

            }
            catch (WebException e)
            {
                return null;
            }
            responseTask = JsonConvert.DeserializeObject<dto.Task>(result);

            return responseTask;
        }

        public List<Category> findCategoriesByUserID()
        {

            List<Category> categories = null;

            System.Net.HttpWebRequest webrequest = (HttpWebRequest)System.Net.WebRequest.Create(URL_BASE + "/users/" + userId + "/categories");
            webrequest.Method = "GET";
            webrequest.ContentType = "application/json";

            webrequest.Headers.Add("Authorization", encodedCredentials);

            string result;

            try
            {
                using (WebResponse response = webrequest.GetResponse())
                {
                    using (StreamReader reader = new StreamReader(response.GetResponseStream()))
                    {
                        result = reader.ReadToEnd();
                    }
                }

            }
            catch (WebException e)
            {
                return null;
            }


            categories = JsonConvert.DeserializeObject<List<Category>>(result);

            return categories;
        }

        public void markTaskAsFinished(long id)
        {

            System.Net.HttpWebRequest webrequest = (HttpWebRequest)System.Net.WebRequest.Create(URL_BASE + "/users/" + userId + "/tasks/" + id);
            webrequest.Method = "PUT";
            webrequest.ContentType = "application/json";

            webrequest.Headers.Add("Authorization", encodedCredentials);

            try
            {
                WebResponse response = webrequest.GetResponse();
                //if(response.)
            }
            catch (WebException e)
            {
                System.Console.WriteLine("No se ha podido marcar la tarea");
            }

        }

        public List<dto.Task> getTasksFromCategoryId(long id)
        {

            List<dto.Task> categories = null;

            System.Net.HttpWebRequest webrequest = (HttpWebRequest)System.Net.WebRequest.Create(URL_BASE + "/users/" + userId + "/categories/" + id + "/tasks");
            webrequest.Method = "GET";
            webrequest.ContentType = "application/json";

            webrequest.Headers.Add("Authorization", encodedCredentials);

            string result;

            try
            {
                using (WebResponse response = webrequest.GetResponse())
                {
                    using (StreamReader reader = new StreamReader(response.GetResponseStream()))
                    {
                        result = reader.ReadToEnd();
                    }
                }

            }
            catch (WebException e)
            {
                return null;
            }


            categories = JsonConvert.DeserializeObject<List<dto.Task>>(result);

            return categories;
        }
    }
}
